Replace more usages of Type with DeprecatedTypeProperties

li-roy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#19093 Replace more usages of Type with DeprecatedTypeProperties**

- Made at::CPU(), at::CUDA() return DeprecatedTypeProperties
- To better support bc, added more functions to DeprecatedTypeProperties
- Added is_variable_ field to DeprecatedTypeProperties to better mirror old Type behavior

Differential Revision: [D14865520](https://our.internmc.facebook.com/intern/diff/D14865520)

Commit:
Replace more usages of Type with DeprecatedTypeProperties

ezyang commented on aten/src/ATen/core/DeprecatedTypeProperties.h:
```diff
@@ -41,8 +50,8 @@ class DeprecatedTypeProperties {
-  bool is_defined() const {
-    return backend_ != Backend::Undefined && scalar_type_ != ScalarType::Undefined;
+  bool is_variable() const {
+    return is_variable_;
```
What happened here?

li-roy replied:
is_defined() wasn't something on Type before, I was using it as a helper but looks like I actually don't need it.

is_variable() is something I want to move over.

ezyang commented on aten/src/ATen/core/DeprecatedTypePropertiesRegistry.h:
```diff
@@ -5,40 +5,29 @@
 class CAFFE2_API DeprecatedTypePropertiesRegistry {
  public:
   using DeprecatedTypePropertiesUniquePtr =
       std::unique_ptr<DeprecatedTypeProperties, DeprecatedTypePropertiesDeleter>;
```
Reminding myself: the reason the deleter is explicitly specified here, is because we only have a forward declaration of DeprecatedTypeProperties at this point (so the default deleter won't work); is that right?

li-roy replied:
That sounds right. But now that you mention it I actually don't need it anymore because I moved definitions to the cpp file.

ezyang commented on torch/csrc/Generator.cpp:
```diff
@@ -82,8 +82,7 @@ static PyObject * THPGenerator_setState(THPGenerator *self, PyObject *_new_state
   auto& tensor = ((THPVariable*)_new_state)->cdata.data();
-  auto& tensor_type = at::globalContext().getNonVariableType(tensor.type().backend(), tensor.scalar_type());
-  if (tensor_type != CPU(kByte)) {
+  if (tensor.type().backend() != Backend::CPU || tensor.scalar_type() != kByte) {
```
`tensor.layout() == kStrided && tensor.device().type() == kCPU` instead?

ezyang commented on aten/src/ATen/templates/TensorMethods.h:
```diff
@@ -6,37 +6,41 @@
     return *this;
-  return t.copy(*this, non_blocking);
+  return to(
+      at::device(t.device_type()).layout(t.layout()).dtype(t.scalarType()),
```
Wouldn't `t.dtype()` be better here?

li-roy replied:
It's the same because t is a Type which stores a ScalarType. we do exactly one conversion in both cases.

ezyang commented on aten/src/ATen/templates/Tensor.h:
```diff
@@ -199,7 +200,9 @@ class CAFFE2_API Tensor {
-        tensorTypeIdToBackend(type_id()), scalar_type());
+        tensorTypeIdToBackend(type_id()),
+        scalar_type(),
+        is_variable() && !at::NonVariableTypeMode::is_enabled());
```
I'm not sure how to tell that you needed to check the variable type mode here. (As in, I could believe that this change is needed, but I don't know how to show it's correct.)

li-roy replied:
I don't know if it's required, but I just want to align it with the way Type is picked for now.

ezyang reviewed (approved):
Approved with... shall I say, *inquiries* ;)

gchanan commented on aten/src/ATen/core/DeprecatedTypeProperties.cpp:
```diff
@@ -0,0 +1,24 @@
+
+namespace at {
+
+Tensor DeprecatedTypeProperties::unsafeTensorFromTH(void * th_pointer, bool retain) const {
```
how did you gather this collection of functions to add?

gchanan replied:
also, is the plan here that the implementations of these functions will move here?  Because if dispatch type loses scalars, the implementations probably won't work as is, right?

li-roy replied:
I think the only function that people would actually use off of Type is copy. The others I moved because the tests access them from at::CPU(), which now returns DeprecatedTypeProperties.

I'm going to keep these calls here for now. Copy I'm moving to native, and i'll just call an equivalent here instead. The others don't actually need to come off of Type right away because they don't use dtype.

Commit:
Update on "Replace more usages of Type with DeprecatedTypeProperties"

Replace more usages of Type with DeprecatedTypeProperties

gh-metadata: pytorch pytorch 19093 gh/li-roy/5/head

Commit:
Update on "Replace more usages of Type with DeprecatedTypeProperties"

Replace more usages of Type with DeprecatedTypeProperties

gh-metadata: pytorch pytorch 19093 gh/li-roy/5/head

Commit:
Update on "Replace more usages of Type with DeprecatedTypeProperties"

Replace more usages of Type with DeprecatedTypeProperties

gh-metadata: pytorch pytorch 19093 gh/li-roy/5/head
